lib/pbench/server/database/models/server_config.py

```diff
@@ -0,0 +1,389 @@
+    """
+    Attempt to set a ServerConfig with an empty key.
+    """
+
+    def __str__(self) -> str:
+        return "Missing key name"
+
+
+class ServerConfigBadKey(ServerConfigError):
+    """
+    Attempt to commit a ServerConfig with a bad key.
+    """
+
+    def __init__(self, key: str):
+        self.key = key
+
+    def __str__(self) -> str:
+        return f"Configuration key {self.key!r} is unknown"
+
+
+class ServerConfigBadValue(ServerConfigError):
+    """
+    Attempt to assign a bad value to a server configuration option
+    """
+
+    def __init__(self, key: str, value: JSONVALUE):
+        self.key = key
+        self.value = value
+
+    def __str__(self) -> str:
+        return f"Unsupported value for configuration key {self.key!r} ({self.value!r})"
+
+
+# Formal timedelta syntax is "[D day[s], ][H]H:MM:SS[.UUUUUU]"; without an
+# external package we have no standard way to parse or format timedelta
+# strings. Additionally for "lifetime" we don't care about fractional days,
+# so we simpify by accepting "D[[ ]day[s]]".
+_TIMEDELTA_FORMAT = re.compile(r"(?P<days>\d{1,9})(?:\s*day(?:s)?)?")
+
+
+def validate_lifetime(key: str, value: JSONVALUE) -> JSONVALUE:
+    v = str(value)
+    check = _TIMEDELTA_FORMAT.fullmatch(v)
+    if check:
+        days = check.group("days")
+    else:
+        raise ServerConfigBadValue(key, value)
+    return days
+
+
+# Formal "state" syntax is a JSON object with a "status" key designating the
+# current server status ("enabled", "disabled", or "readonly" to allow read
+# access but not modification of resources), and a "message" string
+# describing why the server isn't enabled. The "message" is required only if
+# the "status" isn't "enabled". For example,
+#
+# {"status": "enabled"}
+#
+# or
+#
+# {"status": "disabled", "message": "Down for maintanance"}
+#
+# or
+#
+# {"status": "readonly", "message": "Available for queries but no changes allowed"}
+#
+# Additional fields can be set by an administrator and will be retained and
+# reported when the state is queried or when an API call fails because the
+# state isn't enabled.
+STATE_STATUS_KEY = "status"
+STATE_MESSAGE_KEY = "message"
+STATE_STATUS_KEYWORD_DISABLED = "disabled"
+STATE_STATUS_KEYWORD_ENABLED = "enabled"
+STATE_STATUS_KEYWORD_READONLY = "readonly"
+STATE_STATUS_KEYWORDS = [
+    STATE_STATUS_KEYWORD_DISABLED,
+    STATE_STATUS_KEYWORD_ENABLED,
+    STATE_STATUS_KEYWORD_READONLY,
+]
+
+
+def validate_server_state(key: str, value: JSONVALUE) -> JSONVALUE:
+    try:
+        status = value[STATE_STATUS_KEY].lower()
+    except (KeyError, SyntaxError, TypeError):
+        raise ServerConfigBadValue(key, value)
+    if status not in STATE_STATUS_KEYWORDS:
+        raise ServerConfigBadValue(key, value)
+
+    # canonicalize the status value by lowercasing it
+    value[STATE_STATUS_KEY] = status
+    if status != STATE_STATUS_KEYWORD_ENABLED and STATE_MESSAGE_KEY not in value:
+        raise ServerConfigBadValue(key, value)
```

Comment:
I don't see any point to a message for an enabled state; it'd only be visible if someone asked for it explicitly, and that's really what the banner is for. However, I don't fail if you choose to specify a message.

I'd rather not fabricate some inane default disabled message, and I don't see any possible value in the extra code. I could certainly make it optional, but why?

Reply:
> that's really what the banner is for

I was initially going to point out that we could collapse the two options and store the banner with the `enabled` state, but then I concluded that the banner might be something that we wanted displayed independent of the server state.  But, under that banner, we might have a secondary display which _is_ state-driven.  And, so, being able to have a `"message"` regardless of whether the server is enabled or not might be handy (depending on what the client wants to do).

Correspondingly, having a message to display when the server is only in certain states seems somewhat weird/awkward/cumbersome.  I think it would simplify the code, especially on the client side, if the presumption were that there is a message, regardless of the state (although, the client is free to ignore it...).

As for fabricating some inane default messages, I'm willing to do that for you....  😆
- for `"disabled"`, we can go with, "The server is currently unavailable, please try again later."
- for `"readonly"`, we can go with, "The server is currently available in a read-only mode, uploads and other changes are not currently available."  And,
- for `"enabled"`, how about, "Welcome to the Pbench Server, home for all your data!"

Or, we can just use empty strings.  (The problem with human-readable messages is, what language should we be composing them in??)

I recommend making the messages optional since there's no guarantee that the client will display them to the user.  The reason for making them optional is that (I think) it will make the server-side code simpler (e.g., we don't have to have the code that I've cited here).  Likewise, I think that having defaults will make things simpler as well, because various parts of the code can assume that there is a message, because the default message will be there if the client didn't supply one, and the client's life is simpler because they only specify a message if they want to.